src/core/thread/mle_router.cpp

```diff
@@ -4483,6 +4484,59 @@ Error MleRouter::SendTimeSync(void)
 }
 #endif // OPENTHREAD_CONFIG_TIME_SYNC_ENABLE
 
+Error MleRouter::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    if (IsChild())
+    {
+        ExitNow(error = Mle::DetachGracefully(aCallback, aContext));
+    }
```

Comment:
What if `DetachGracefully()` is called while we are detached  (`kRoleDetached`) or disabled (`kRoleDisabled`)?
Maybe we can return `kErrorInvalidState`?

Reply:
It disables Thread at the next frame and calls the callback, the same as when being a leader.

I think it's better to return OK in this case, to be consistent with that calling "thread stop" while in disabled state does nothing and returns OK.